### ISSUE TITLE
Partial incremental tasks with > 1 (discovered) input per output

### ISSUE THREAD

**Issue by ranweiler** [nightroman/Invoke-Build#141]
## Problem

I want to define a task _like_ a partial incremental task, but where one output depends on > 1 input file, and the input file groups are discovered at task invocation time.

I don't think the existing API directly supports this, but I could be mistaken.

## Toy example

Suppose I have a directory like this:

```
lib1/
  ./a.obj
  ./b.obj
lib2/
  ./c.obj
# ...
libN/
  ./x.obj
  ./y.obj
  ./z.obj
```

My goal is to incrementally build `N` DLLs, named `lib1.dll`, `lib2.dll`, ..., `libN.dll`. If any `.obj` file is updated after its parent `.dll` has been built, the `.dll` should be rebuilt.

Constraints:
- I don't want to define an explicit `task` for each DLL: I have too many.
- I really don't want `N` to be fixed by the build script: I want to discover it.
- I can't use partial incremental tasks because I map _many_ inputs to 1 output.
- I want building to be efficient, i.e. incremental _per_-DLL.

## Details

In my specific situation, each output is determined by a _set_ of input files. There will be _many_ sets, and I want to discover the input file sets at task invocation time.

Incremental tasks partly support my use case, because each task describes how to use a _set_ of input files to get a set of output files (or just 1 output file). The downside is that we need one task defined, statically, for each input/output pair. Incremental tasks seem to support my cardinality needs, but are insufficiently dynamic.

OTOH, with _partial_ incremental tasks, I can _dynamically_ discover a set of input files, and then give a "pointwise rule" to produce exactly one output file from each input file _in the set_. However, the restricted input file count (exactly 1) is the sticking point for my use case. In other scenarios, I could imagine the restricted output file count (exactly 1) being a sticking point.

## Solutions 

In the Invoke-Build API, a partial incremental task is concretely defined as one task, but is in some sense _many_ tasks, where the number of tasks may be dynamic. Alternately, we can think of a partial incremental task as a single _parametric_ task that operates on a set of inputs.

One solution could be generalizing the existing partial incremental task API. Instead of `$Input` and `$Output` being equal-length arrays of _files_, they could be equal length arrays of _collections_ of files.

Alternately, there could be some notion of an explicitly parametric incremental task that accepts `$Inputs` and `$Outputs` as arguments. Then, an invoking task could discover the `$Inputs` array (whose elements may themselves be arrays) at runtime, define expected `$Output`, and invoke the parametric task, which would be defined as a the full N-M incremental task.

If I've overlooked some way to accomplish my goal without extending the existing API, that's even better.

**Comment by nightroman** [comment]
Thank you for the well described and very practical problem.
The below script demonstrates the technique you can use.

Original constraints / solutions:

* I don't want to define an explicit `task` for each DLL: I have too many.
    * We have many tasks but we generate them, not code each manually
* I really don't want `N` to be fixed by the build script: I want to discover it.
    * N is discovered (it's the number of folders in the parent folder)
* I can't use partial incremental tasks because I map _many_ inputs to 1 output.
    * We do not need partial incremental tasks with this approach.
* I want building to be efficient, i.e. incremental _per_-DLL.
    * It is incremental (per "z.X.log" in the sample)
* I want to discover the input file sets at task invocation time.
    * All the job is done only when `ManyIncrementalTasks` is called.

Is it a sufficient solution or do we need some extra helper API?
So far I do not think that some extra API will really help much.
Discovery of inputs/outputs will have to be coded anyway.

```powershell
task ManyIncrementalTasks {
    # list of data sets for generated tasks
    $taskData = [System.Collections.Generic.List[object]]::new()

    # collect data sets for each directory in the parent directory
    foreach($dir in Get-ChildItem .. -Directory) {
        # one input set and its output file name
        $in = Get-ChildItem -LiteralPath $dir.FullName -Recurse -File
        $out = "z.$($dir.Name).log"

        # parameters of one incremental task
        # - here we use $out as the task name
        # - the job is to write input names to "z.X.log"
        $taskData.Add(@{
            Name = $out
            Inputs = $in
            Outputs = $out
            Jobs = {
                $Inputs | Set-Content $Outputs
            }
        })
    }

    # invoke all tasks (*) in a dynamic script with dynamic incremental tasks
    Invoke-Build * {
        foreach($data in $taskData) {
            task @data
        }
    }
}

task Clean {
    remove z.*
}
```

**Comment by nightroman** [comment]
Here is some shorter version

```powershell
# invoke all tasks (*) in a dynamic script with dynamic incremental tasks
task ManyIncrementalTasks {
    Invoke-Build * {
        # collect data sets for each directory in the parent directory
        foreach($dir in Get-ChildItem .. -Directory) {
            # one input set and its output file name
            $in = Get-ChildItem -LiteralPath $dir.FullName -Recurse -File
            $out = "z.$($dir.Name).log"

            # parameters of one incremental task
            # - here we use $out as the task name
            # - the job is to write input names to "z.X.log"
            task $out @{
                Inputs = $in
                Outputs = $out
                Jobs = {
                    $Inputs | Set-Content $Outputs
                }
            }
        }
    }
}
```

**Comment by ranweiler** [comment]
@nightroman, your approach seems to be exactly what I need. Being able to dynamically generate incremental tasks is really the best-case solution for my actual problem.

The only wrinkle I encountered was that this forces the user to reckon with various scoping subtleties (e.g. a custom `$BuildRoot` isn't automatically propagated to the dynamically-generated tasks), but those arise from PowerShell semantics, not any problem with Invoke-Build.

Closing this issue as resolved. Thank you!
